vowpalwabbit/cb_dro.cc

```diff
@@ -0,0 +1,154 @@
+#include "reductions.h"
+#include "cb_dro.h"
+#include "distributionally_robust.h"
+#include "explore.h"
+
+#include "rand48.h"
+
+using namespace LEARNER;
+using namespace VW;
+using namespace VW::config;
+
+namespace VW
+{
+struct cb_dro_data
+{
+  explicit cb_dro_data(double alpha, double tau, double wmax) : chisq(alpha, tau, 0, wmax) {}
+
+  template <bool is_learn, bool is_explore>
+  inline void learn_or_predict(multi_learner &base, multi_ex &examples)
+  {
+    // Some explanation required.
+    //
+    // VW currently is either
+    //    in prediction mode with exploration (aka --cb_explore_adf)
+    //    or in (off-policy) learning without exploration (aka --cb_adf)
+    //
+    // In practice, nobody seems to do
+    // "off policy learning of the stochastic exploration policy".
+    //
+    // Ergo, the following always optimizes the bound on the associated
+    // deterministic argmax score policy.
+
+    multiline_learn_or_predict<false>(base, examples, examples[0]->ft_offset);
+
+    if (is_learn)
+    {
+      auto it = std::find_if(examples.begin(), examples.end(), [](example *item) { return !item->l.cb.costs.empty(); });
+
+      if (it != examples.end())
+      {
+        CB::cb_class logged = (*it)->l.cb.costs[0];
+        uint32_t labelled_action = std::distance(examples.begin(), it);
+
+        auto action_scores = examples[0]->pred.a_s;
+
+        // cb_explore_adf => want maximum probability
+        // cb_adf => want minimum cost
+
+        auto maxit = is_explore
+                     ? std::max_element (action_scores.begin(),
+                                         action_scores.end(),
+                                         [](const ACTION_SCORE::action_score& a, const ACTION_SCORE::action_score& b) { return ACTION_SCORE::score_comp(&a, &b) < 0; })
+                     : std::min_element (action_scores.begin(),
+                                         action_scores.end(),
+                                         [](const ACTION_SCORE::action_score& a, const ACTION_SCORE::action_score& b) { return ACTION_SCORE::score_comp(&a, &b) < 0; });
+        uint32_t chosen_action = maxit->action;
+
+        float w = logged.probability > 0 ? 1 / logged.probability : 0;
+        float r = -logged.cost;
+
+        chisq.update(chosen_action == labelled_action ? w : 0, r);
+
+        float qlb = chisq.effn() * chisq.qlb(w, r);
+
+        // avoid pathological cases
+        qlb = std::max(qlb, 0.01f);
+
+        // save the original weights and scale the example weights
+        std::vector<float> save_weight;
```

Comment:
You can reserve so that there aren't several allocation as it grows

Reply:
Done.

Reply:
If save_weight is made a member variable of cb_dro_data, and clear()ed here.  Allocations in steady state should go to zero in predict_learn().